src/GafferUI/Gadget.cpp

```diff
@@ -56,6 +56,54 @@ using namespace std;
 
 GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( Gadget );
 
+//////////////////////////////////////////////////////////////////////////
+// Gadget::Signals
+//
+// We allocate these lazily because they have a significant overhead
+// in both memory and construction time, and for many Gadget
+// instances they are never actually used.
+//////////////////////////////////////////////////////////////////////////
+
+struct Gadget::Signals : boost::noncopyable
+{
+
+	VisibilityChangedSignal visibilityChangedSignal;
+	RenderRequestSignal renderRequestSignal;
+
+	ButtonSignal buttonPressSignal;
+	ButtonSignal buttonReleaseSignal;
+	ButtonSignal buttonDoubleClickSignal;
+	ButtonSignal wheelSignal;
+
+	EnterLeaveSignal enterSignal;
+	EnterLeaveSignal leaveSignal;
+	ButtonSignal mouseMoveSignal;
+
+	DragBeginSignal dragBeginSignal;
+	DragDropSignal dragEnterSignal;
+	DragDropSignal dragMoveSignal;
+	DragDropSignal dragLeaveSignal;
+	DragDropSignal dragEndSignal;
+	DragDropSignal dropSignal;
+
+	KeySignal keyPressSignal;
+	KeySignal keyReleaseSignal;
+
+	// Utility to emit a signal if it has been created, but do nothing
+	// if it hasn't.
+	template<typename SignalMemberPointer, typename... Args>
+	static void emitLazily( Signals *signals, SignalMemberPointer signalMemberPointer, Args&&... args )
```

Comment:
I'm not sure on `emitLazily` as the name here. We have `GafferUI.LazyMethod` which is explicitly about deferring/de-duping calls - which I consequently thought this was for too when I saw it at the call sites. As I understand it, this is just a `no-op` if no signals exist yet. Just `emit` doesn't seem too bad? I couldn't think of anything better with more words in that wasn't misleading in one way or another...

Reply:
I see your point, but I was just following the idiom already established in GraphComponent. Want me to rename it all?

Reply:
Ahh, got you. If we already have this ambiguity, then happy to leave as is, unless you really want to change both...

Reply:
I'm pretty happy knowing that laziness can mean different things in different contexts. If we really want to get into renaming then LazyMethod->DeferredMethod may be a better candidate - it being in the public API means it actually matters, and deferred is a bit more descriptive than lazy.

Reply:
Yeah, with you on that. Leave as is for now then :)